migration/ubuntu/20.04-to-22.04-openstack.md

```diff
@@ -0,0 +1,256 @@
+> - If it fails late for some with timeouts then ignore.
+
+- Check that old etcd members are removed:
+
+  ```bash
+  HOSTNAME="$(kubectl get node -l node-role.kubernetes.io/control-plane= -ojsonpath='{.items[0].metadata.name}')"
+  kubectl -n kube-system exec -it "etcd-${HOSTNAME}" -- \
+    etcdctl --endpoints=https://127.0.0.1:2379 \
+            --cacert=/etc/kubernetes/ssl/etcd/ca.crt \
+            --cert=/etc/kubernetes/ssl/etcd/server.crt \
+            --key=/etc/kubernetes/ssl/etcd/server.key \
+            member list
+  ```
+
+- If there are old nodes left, remove them using the `uuid` from the previous command (The id in the first column of each row)
+
+  ```bash
+  HOSTNAME="$(kubectl get node -l node-role.kubernetes.io/control-plane= -ojsonpath='{.items[0].metadata.name}')"
+  kubectl -n kube-system exec -it "etcd-${HOSTNAME}" -- \
+    etcdctl --endpoints=https://127.0.0.1:2379 \
+            --cacert=/etc/kubernetes/ssl/etcd/ca.crt \
+            --cert=/etc/kubernetes/ssl/etcd/server.crt \
+            --key=/etc/kubernetes/ssl/etcd/server.key \
+            member remove <uuid>
+  ```
+
+- Check that old control plane nodes are removed:
+
+  ```bash
+  kubectl get nodes
+  ```
+
+- If there are any old nodes left, delete them
+
+  ```bash
+  kubectl delete nodes <node>,...
+  ```
+
+- Change back to using dynamic inventory
+
+  ```bash
+  pushd "${CK8S_CONFIG_PATH}"
+  cp <sc|wc>-config-backup/inventory.ini <sc|wc>-config/inventory.ini
+  popd
+  ```
+
+- Remove old control plane nodes using Terraform
+- Remove the old control plane floating IPs from group vars under the key `supplementary_addresses_in_ssl_keys`
+
+> [!NOTE]
+> This is not needed if you are using a loadbalancer in front of the API server.
+
+- Apply kubespray:
+
+  ```bash
+  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
+  ck8s-kubespray apply <sc|wc> --tags=master,nginx -b
+  popd
+  ```
+
+- Update the NetworkPolicies:
+
+  ```bash
+  pushd "${CK8S_APPS_REPOSITORY_PATH}"
+  ck8s update-ips <sc|wc> apply
+
+  ck8s ops helmfile <sc|wc> -l app=netpol -i apply
+  popd
+  ```
+
+### Replace worker nodes
+
+- Add a new worker node that doesn't specify `image_id` (To use the new image)
+- Verify by running `terraform plan` and make sure there's a new node added with the correct image
+- Add the rest of the new worker nodes
+- Run `terraform apply` to add all the new nodes
+- Add all the new worker nodes
+
+  ```bash
+  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
+  ck8s-kubespray run-playbook <sc|wc> facts.yml
+  ck8s-kubespray run-playbook <sc|wc> scale.yml -b --limit=<new-node1>,<new-node2>,...
+  popd
+  ```
+
+- Drain the old nodes and make sure everything starts up properly
+
+> [!CAUTION]
+> If you are using applications that allowlist the old nodes, please update that list to also include the new nodes before draining the old ones.
+
+  ```bash
+  kubectl drain <old-node1>
+  kubectl drain <old-node2>
+  ...
```

Comment:
If this is meant to drain the nodes one at a time, then could you add a step here to first cordon all old nodes that will be removed? That way you don't have pods getting drained from one old node to another old node, just to get drained again later.

Reply:
https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/5feea8586cdaf1dab62242923751189edbce4293